From 0cff2d6da11944afc61977b9c8a554b4edd5a089 Mon Sep 17 00:00:00 2001
From: Mathieu Bivert <mathieu.bivert@intel.com>
Date: Wed, 20 Mar 2013 17:21:17 +0100
Subject: [PORT FROM MAIN] wifi/crashtool: log connection time for brcm supplicant

BZ: 85467

Add trace for connection time.

Change-Id: I331e84fdbd366c080ac6646d7e024a2ed75ee8f1
Orig-Change-Id: I5085303598e5c1d7dc88b759bedfbe0685cbed58
Category: device enablement
Domain: CWS.WIFI-Common
Origin: upstream
Upstream-Candidate: no,proprietary
Signed-off-by: Axel Haslam <axelx.haslam@intel.com>
Signed-off-by: Mathieu Bivert <mathieu.bivert@intel.com>
Reviewed-on: http://android.intel.com:8080/122732
Reviewed-by: buildbot <buildbot@intel.com>
Reviewed-by: Fiat, Christophe <christophe.fiat@intel.com>
Tested-by: Champciaux, Nicolas <nicolas.champciaux@intel.com>
Reviewed-by: cactus <cactus@intel.com>
Tested-by: cactus <cactus@intel.com>
---
 wpa_supplicant/wpa_supplicant.c |   44 +++++++++++++++++++++++++++++++++++++++
 1 files changed, 44 insertions(+), 0 deletions(-)

diff --git a/wpa_supplicant/wpa_supplicant.c b/wpa_supplicant/wpa_supplicant.c
index 7ccf273..a690029 100644
--- a/wpa_supplicant/wpa_supplicant.c
+++ b/wpa_supplicant/wpa_supplicant.c
@@ -52,6 +52,7 @@
 #include "hs20_supplicant.h"
 #include "wnm_sta.h"
 #include "bestchannel.h"
+#include <lct.h>
 
 const char *wpa_supplicant_version =
 "wpa_supplicant v" VERSION_STR "\n"
@@ -611,6 +612,47 @@ static void wpa_supplicant_stop_bgscan(struct wpa_supplicant *wpa_s)
 
 #endif /* CONFIG_BGSCAN */
 
+static unsigned long timevaldiff(struct timeval *starttime, struct timeval *finishtime)
+{
+	unsigned long msec;
+	msec = (finishtime->tv_sec - starttime->tv_sec) * 1000;
+	msec += (finishtime->tv_usec - starttime->tv_usec) / 1000;
+	return msec;
+}
+
+static inline void log_connection_time(struct wpa_supplicant *wpa_s,
+                                      enum wpa_states state)
+{
+	struct ct_event *ev = NULL;
+	static struct timeval start = {
+		.tv_sec = 0,
+		.tv_usec = 0
+	};
+	struct timeval stop;
+	unsigned long msec_to_connect = 0;
+	char msec_c[32];
+	/* Crashtool, log time it takes to associate */
+	if (state == WPA_SCANNING &&
+		(wpa_s->wpa_state == WPA_DISCONNECTED ||
+		wpa_s->wpa_state == WPA_INACTIVE)) {
+			gettimeofday(&start, NULL);
+	}
+	else if ((start.tv_sec || start.tv_usec) &&
+		state == WPA_COMPLETED &&
+		(wpa_s->wpa_state == WPA_GROUP_HANDSHAKE  ||
+		wpa_s->wpa_state == WPA_ASSOCIATED)) {
+
+		gettimeofday(&stop, NULL);
+		msec_to_connect = timevaldiff(&start, &stop);
+		bzero(msec_c, sizeof(msec_c));
+		snprintf(msec_c, sizeof(msec_c), "%lu", msec_to_connect);
+		lct_log(CT_EV_STAT, "dev.log.cws.wifi", "conn_time", EV_FLAGS_PRIORITY_LOW, msec_c);
+
+		start.tv_sec = 0;
+		start.tv_usec = 0;
+	}
+}
+
 
 static void wpa_supplicant_start_autoscan(struct wpa_supplicant *wpa_s)
 {
@@ -652,6 +694,8 @@ void wpa_supplicant_set_state(struct wpa_supplicant *wpa_s,
 		wpa_supplicant_state_txt(wpa_s->wpa_state),
 		wpa_supplicant_state_txt(state));
 
+	log_connection_time(wpa_s, state);
+
 #ifdef ANDROID_P2P
 	if(state == WPA_ASSOCIATED && wpa_s->current_ssid) {
 		wpa_s->current_ssid->assoc_retry = 0;
-- 
1.7.4.1

